src/commands/UserUpsertCommand.ts

```diff
@@ -0,0 +1,75 @@
+import {
+  MultisigUserUpdateMessage,
+  USER_UPDATE_MESSAGE_SCHEMA,
+} from "../lib/users/schemas.js";
+import { isTypedMessage } from "../utils/signatures.js";
+import UserUpsertSignatureVerifier from "../lib/safe-signature-verification/UserUpsertSignatureVerifier.js";
+import { Database } from "../types/supabaseData.js";
+
+import { SafeApiCommand } from "./SafeApiCommand.js";
+
+type SignatureRequest =
+  Database["public"]["Tables"]["signature_requests"]["Row"];
+
+export class UserUpsertCommand extends SafeApiCommand {
+  async execute(): Promise<void> {
+    const signatureRequest = await this.dataService.getSignatureRequest(
+      this.safeAddress,
+      this.messageHash,
+    );
+
+    if (!signatureRequest || signatureRequest.status !== "pending") {
+      return;
+    }
+
+    const safeMessage = await this.safeApiKit.getMessage(this.messageHash);
+
+    if (!isTypedMessage(safeMessage.message)) {
+      throw new Error("Unexpected message type: not EIP712TypedData");
+    }
+
+    const message = USER_UPDATE_MESSAGE_SCHEMA.safeParse(
+      safeMessage.message.message,
+    );
+    if (!message.success) {
+      console.log("Unexpected message format", message.error);
+      throw new Error("Unexpected message format");
+    }
+
+    const verifier = new UserUpsertSignatureVerifier(
+      Number(signatureRequest.chain_id),
+      this.safeAddress as `0x${string}`,
```

Comment:
you can use `getAddress` from `viem` here to verify and not force the type

Reply:
Great point, thank you!